r/R/expression.R

```diff
@@ -217,7 +217,23 @@ build_expr <- function(FUN,
     } else if (FUN == "%/%") {
       # In R, integer division works like floor(float division)
       out <- build_expr("/", args = args)
-      return(out$cast(int32(), allow_float_truncate = TRUE))
+
+      # integer output only for all integer input
+      int_type_ids <- Type[toupper(INTEGER_TYPES)]
+      numerator_is_int <- args[[1]]$type_id() %in% int_type_ids
+      denominator_is_int <- args[[2]]$type_id() %in% int_type_ids
+
+      if (numerator_is_int && denominator_is_int) {
+        out_float <- build_expr(
+          "if_else",
+          build_expr("equal", args[[2]], 0L),
+          Scalar$create(NA_integer_),
+          out
+        )
+        return(out_float$cast(args[[1]]$type(), allow_float_truncate = TRUE))
```

Comment:
🤔 Would it be possible to take out `allow_float_truncate = TRUE` and use `floor()` here to accomplish the truncation bit? When this code was originally written `floor()` didn't exist and we used `allow_float_truncate` to accomplish that (though like you show here that leads to some very silly results!)